### PR TITLE
Add paradise's version of byond-tracy for better production profiling

### PR DESCRIPTION
**Pull request by io-util**
Big credit for actually making [byond-tracy](https://github.com/mafemergency/byond-tracy) goes to mafemergency, additional credit goes to Affectedarc07 which made the modified version that can be found here: [para-byond-tracy](https://github.com/ParadiseSS13/byond-tracy).

This Pull Request does what it says, it adds Paradise Station's version of byond-tracy in the addition to the original byond-tracy, this version of byond-tracy continuously writes a .utracy file of the current round into the data/profiler directory which allows you to 'replay' said file using the replay.py script and analyze it even after a round already ended. 

⚠️ **Fair Warning**: These files get _very_ large. Initialization on the Horizon alone (aka. starting the server) already comes to around a gigabyte, an entire round with no extensions will likely be resulting in a file sitting at around 30~ GB. To replay these .utracy files into Tracy itself you will need a _good_ computer with sufficient memory, I suggest at least 64 GB of memory to profile an entire round.

The client binaries for Tracy itself are shipped into the repo, since byond-tracy requires specifically version 0.8.2.

Both `tracy.dll` and `tracy-disk.dll` have been built by myself, on my windows machine from source using Clang. The Tracy client files are taken directly from the 0.8.2 release on the github page [here](https://github.com/wolfpld/tracy/releases/tag/v0.8.2).

<details><summary>Binary Hashes and VirusTotal links.</summary>
<p>

`tracy.dll`:
File hash is `88499E59D5B2331A51B5AD58B7BBD45F1943C58A5B09FA7B86F372BEDF2C951E`. [VirusTotal](https://www.virustotal.com/gui/file/88499e59d5b2331a51b5ad58b7bbd45f1943c58a5b09fa7b86f372bedf2c951e).

`tracy-disk.dll`:
File hash is `5EFE421D7A2AD1C27E75D3A4422543DA5688F92F5D02574F2DA97C496B0D9100`. [VirusTotal](https://www.virustotal.com/gui/file/5efe421d7a2ad1c27e75d3a4422543da5688f92f5d02574f2da97c496b0d9100).

`Tracy-0.8.2.exe`:
File hash is `DEB5B87CC8FA750443E09F9263D9F6A5E9376E729D6C16727C32EFFC47D0137B`. [VirusTotal](https://www.virustotal.com/gui/file/deb5b87cc8fa750443e09f9263d9f6a5e9376e729d6c16727c32effc47d0137b).

`capture.exe`:
File hash is `773E0905F041ABA1D2C733E97A8B70FDD68F43B69FB8117F69D932DBE506FCFC`. [VirusTotal](https://www.virustotal.com/gui/file/773e0905f041aba1d2c733e97a8b70fdd68f43b69fb8117f69d932dbe506fcfc).

`csvexport.exe`:
File hash is `B190F21771E832FE6F8183BCE779D584D0699758A3501FD5D4129E721F4C6A2B`. [VirusTotal](https://www.virustotal.com/gui/file/b190f21771e832fe6f8183bce779d584d0699758a3501fd5d4129e721f4c6a2b).

</p>
</details> 

Some AV's I never heard of on VirusTotal flag some files, likely because they're writing etc. every sane AV ignores them, as they are not malicious. I can ensure you the dll's have been built from source by me and the tracy client files taken from their github releases.

Everything here has been tested by me on BYOND Version 514.1589, but should be compatible with every byond version documented on Paradise's byond-tracy repository.

DNM until Maintainers checked the binaries.

### CODE REVIEW

**Comment by io-util** [comment]
!review

**Comment by io-util** [comment]
Also, if for some reason someone decides to host the code on Linux and want to profile there, you have to build the required libprof.so yourself.

**Comment by FluffyGhoster** [comment]
I would rather this be a separate DLL and tools, as the usage of tracy is pretty useful on the fly during development and losing the realtime-ness (and being stuck on an older version) isn't ideal

**Comment by io-util** [comment]
> I would rather this be a separate DLL and tools, as the usage of tracy is pretty useful on the fly during development and losing the realtime-ness (and being stuck on an older version) isn't ideal

Neither of the both are meant to be in any production environment to begin with, though this version is much easier on server performance than the normal byond-tracy and better suited for use in a production environment. It allows the memory usage of dream daemon to stay relatively low at runtime and is just a little heavier on I/O, and of course, the storage it needs later on, whereas the use of the normal byond-tracy gets heavy on the runtime memory usage of dreamdaemon.

**Comment by io-util** [comment]
Though if the Maintainers wish to have both version in, which would likely cause confusion and in my eyes, is really obsolete, I can happily add the normal byond-tracy.

**Comment by FluffyGhoster** [comment]
> > I would rather this be a separate DLL and tools, as the usage of tracy is pretty useful on the fly during development and losing the realtime-ness (and being stuck on an older version) isn't ideal
> 
> Neither of the two are meant to be in any production environment to begin with, though this version is much easier on server performance than the normal byond-tracy and better suited for use in a production environment. It allows the memory usage of dreamdaemon to stay relatively low at runtime and is just a little heavier on I/O, and of course, the storage it needs later on, whereas the use of the normal byond-tracy gets heavy on the runtime memory usage of dreamdaemon.
> 
> Additionally, it makes it much easier for multiple people to be analyzing the profiles, and of course the best part allowing the storage of profiles as they are all written to files. There is really no reason in my eyes to have to rely on on the normal byond-tracy at this point. Nobody is going to start tracy, connect and then just quickly analyze and patch things as the round is still in progress, the task of analyzing the profiles and then changing things based on observations is a much more post-round focused task.

I have specified "during development", tracy is a good tool to reconstruct bugs and call chains, and evaluate how heavy things are, I have used it extensively in one of my recent PRs and if I had to go through saving/converting/opening/restarting and finding the frames I wanted to look at in a long trace it would have been far less useful, which is why having the realtime connection option is important.

It is also not any more or less heavy on DD, as seen in the PR when I implemented it, the RAM usage is almost exactly the same, the client process the trace, the DLL is pretty much zero cost.

**Comment by io-util** [comment]
> If it's really needed for you to have this feature for local development, you can quickly implement it on a local branch for yourself, I don't see the use of it for a production environment.

Is there an issue that would prevent you from doing _this_ and warrant adding it to the production code?

**Comment by FluffyGhoster** [comment]
> > If it's really needed for you to have this feature for local development, you can quickly implement it on a local branch for yourself, I don't see the use of it for a production environment.
> 
> Is there an issue that would prevent you from doing _this_ and warrant adding it to the production code?

No, it's a matter of convenience, just as nothing prevents you on the production server to pull the branch and swap the DLL underneath, so it's a fairly arbitrary point.

Renaming this new DLL something as "perf-disk.dll" and having it loaded by the config flag, while the verb loads the normal version, is trivial and without any downside, which is what I'm asking to do.

**Comment by io-util** [comment]
> No,

Good 😃

**Comment by FluffyGhoster** [comment]
>I could make the same point that you just made for you

Yes, that's why I have called it arbitrary

>just that in that case something that is not used in production would not be in the code, you just have to swap out a DLL locally.

We have many things in the repo for convenience that are not used in production, this PR is fine code wise, I will just follow it up with another PR by readding the live option if it's merged as is, no big deal either way

**Comment by io-util** [comment]
> having it loaded by the config flag, while the verb loads the normal version

If we are going to talk about convenience then, this is a bad idea. What I would do if implementing it, let the configuration option always use the byond-tracy version that writes to disk and let you choose which version to use when using the verb. I can see moments where I would much rather quickly start tracy, have it write a file to disk and then stop the feature - for example if I want to profile something and then put it in my backlog to work. 

TL;DR - Config option always loads para-byond-tracy, adminverb lets you choose.

**Comment by FluffyGhoster** [comment]
> TL;DR - Config option always loads para-byond-tracy, adminverb lets you choose.

That's exactly what I was thinking and what I said: "perf-disk.dll" and having it loaded by the config flag while the verb loads the normal version

Perf-disk.dll is the DLL that saves to the disk, the normal DLL is what comes from the unmodified repo and does not save on disk

**Comment by io-util** [comment]
Updated everything the way Fluffy asked for it. Edited the title and description, the description includes File Hashes and Virustotal links for every binary file I added for verification.

**Comment by io-util** [comment]
> Also, if for some reason someone decides to host the code on Linux and wants to profile there, you have to build the required libprof.so yourself, I did not do that considering I myself am on a Windows system and Aurora is also hosted on one.

Re: Linux:

If you really have the intention to run Aurora on Linux and use the profiler there, please implement it yourself.

**Review by FluffyGhoster — APPROVED** [review]

**Review by Arrow768 — APPROVED** [review]

**Review by NonQueueingMatt — APPROVED** [review]
